test/compliance/password-grant-type_test.js

```diff
@@ -0,0 +1,233 @@
+/**
+ * Request
+ * @see https://datatracker.ietf.org/doc/html/rfc6749#section-4.3.2
+ *
+ * grant_type
+ *      REQUIRED.  Value MUST be set to "password".
+ * username
+ *      REQUIRED.  The resource owner username.
+ * password
+ *      REQUIRED.  The resource owner password.
+ * scope
+ *      OPTIONAL.  The scope of the access request as described by Section 3.3.
+ */
+
+/**
+ * Response
+ * @see https://datatracker.ietf.org/doc/html/rfc6749#section-5.1
+ *
+ *   access_token
+ *         REQUIRED.  The access token issued by the authorization server.
+ *   token_type
+ *         REQUIRED.  The type of the token issued as described in
+ *         Section 7.1.  Value is case insensitive.
+ *   expires_in
+ *         RECOMMENDED.  The lifetime in seconds of the access token.  For
+ *         example, the value "3600" denotes that the access token will
+ *         expire in one hour from the time the response was generated.
+ *         If omitted, the authorization server SHOULD provide the
+ *         expiration time via other means or document the default value.
+ *   refresh_token
+ *         OPTIONAL.  The refresh token, which can be used to obtain new
+ *         access tokens using the same authorization grant as described
+ *         in Section 6.
+ *   scope
+ *         OPTIONAL, if identical to the scope requested by the client;
+ *         otherwise, REQUIRED.  The scope of the access token as
+ *         described by Section 3.3.
+ */
+
+/**
+ * Response (error)
+ * @see https://datatracker.ietf.org/doc/html/rfc6749#section-5.2
+ *
+ *   error
+ *         REQUIRED.  A single ASCII [USASCII] error code from the following:
+ *         invalid_request, invalid_client, invalid_grant
+ *         unauthorized_client, unsupported_grant_type, invalid_scope
+ *   error_description
+ *         OPTIONAL.  Human-readable ASCII [USASCII] text providing
+ *         additional information, used to assist the client developer in
+ *         understanding the error that occurred.
+ *   error_uri
+ *         OPTIONAL.  A URI identifying a human-readable web page with
+ *         information about the error, used to provide the client
+ *         developer with additional information about the error.
+ */
+
+const OAuth2Server = require('../..');
+const DB = require('../helpers/db');
+const createModel = require('../helpers/model');
+const createRequest = require('../helpers/request');
+const Response = require('../../lib/response');
+
+require('chai').should();
+
+const db = new DB();
+
+const auth = new OAuth2Server({
+  model: createModel(db)
+});
+
+const user = db.saveUser({ id: 1, username: 'test', password: 'test'});
+const client = db.saveClient({ id: 'a', secret: 'b', grants: ['password'] });
+const scope = 'read write';
+
+function createDefaultRequest () {
+  return createRequest({
+    body: {
+      grant_type: 'password',
+      client_id: client.id,
```

Comment:
i think this is "wrong". According to the RFC client_id and secret are not sent via body but via the authorization header with basic auth. Maybe basic auth allows for empty password?

But I also acknowledge that alot of modern oauth implementations make it possible to provide the client_id via the body, I think of okta, auth0 and so...

The problem is the client identification. According to the rfc, there are two types of clients: public and confidential clients. While confidential clients use basicauth to identify themselfs, there is i think no hint how to identify a public client (again: basic auth with empty password?), thus making it impossible to determine the client.

But this is still a non standard request as you quoted yourself in the first lines the expected request for the password grant.

I dont know how to do it better, as i actually also do usually client_id by post for public clients. Maybe I do it wrong and basicauth is the way to go?! But I am 100% sure, that sending the client_secret in the body is 100% wrong, as the spec explicitly defines, that a confidential client has to send the id and password via the authorization header.

Reply:
I think the purpose of these tests is to make sure new code does not brake things. So I'd definitely keep the "wrong" way in these tests, since the library does support this way of authentication. 

I will extend the password grant tests with the other methods for client authorization.

Reply:
@jorenvandeweyer 
Makes sense.

@jankapunkt 
@jwerre 
I guess this is also non conform issue?

Reply:
>   The authorization server MUST support the HTTP Basic
>    authentication scheme for authenticating clients that were issued a
>    client password.

You are right about the basic authentication. So I changed this for both password grant and refresh grant types.

>    Alternatively, the authorization server MAY support including the
>    client credentials in the request-body using the following
>    parameters:

But providing client authentication through the body is actually allowed. So I added `client-authentication_test.js` to test for this compliance.